### DISCUSSION TITLE
Migration to version 4

### DISCUSSION THREAD

**Discussion by marcosflobo** [micronaut-projects/micronaut-core#10822] [Q&A]
Hello team,

Our team found some situations that we believe require some investigations. Hope we can find help here.

Now, let me share some use cases and the solutions/workarounds we found.

# Use case 1: A project dependency has a DTO that requires arbitrary parsing

The following DTO comes from a library dependency.

```java
public class TokenizerResponse {
    @Getter
    public List<Object> parseTree;
}
```

The solution we found: add a mixin
```java
@Serdeable.Deserializable(using = TokenizerResponseDeserializer.class)
public interface TokenizerResponseDeserializationMixin {}
```
```java
package com.xxxxxx.content.manager.externals.datamodel.api.deserialization;
 
import com.xxxxxx.querybuilder.TokenizerResponse;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.core.type.Argument;
import io.micronaut.serde.Decoder;
import io.micronaut.serde.Deserializer;
import jakarta.inject.Singleton;
import java.io.IOException;
import java.util.List;
import lombok.AllArgsConstructor;
 
@Singleton
@AllArgsConstructor
public class TokenizerResponseDeserializer implements Deserializer<TokenizerResponse> {
 
  @Override
  public @Nullable TokenizerResponse deserialize(
      @NonNull Decoder decoder,
      @NonNull DecoderContext context,
      @NonNull Argument<? super TokenizerResponse> type)
      throws IOException {
    var tokens = new TokenizerResponse();
    var rootObjectDecoder = decoder.decodeObject();
    var key = rootObjectDecoder.decodeKey();
    if ("parseTree".equals(key)) {
      tokens.parseTree = parseParseTree(rootObjectDecoder);
    }
    rootObjectDecoder.close();
    return tokens;
  }
 
  private List<Object> parseParseTree(Decoder rootObjectDecoder) throws IOException {
    var result = rootObjectDecoder.decodeArbitrary();
    if (result instanceof List asList) {
      return asList;
    }
    throw new IOException("whatever");
  }
}
```

# Use case 2: custom jackson serializers need to be rewritten using the Micronaut 4 custom serializer
@JsonSerialize using member is not supported yet. So there is no other way than rewrite the serializer

**Solution**
Just rewrite the custom serializers. The API from Micronaut is quite similar to the Jackson API therefore the rewrite is straightforward

# Use case 3: Blocking calls are detected in the event loop and crashes the request
This exception message is thrown by one of our clients that doesn't return reactive constructs

```
"You are trying to run a BlockingHttpClient operation on a netty event loop thread. This is a common cause for bugs: Event loops should never be blocked. You can either mark your controller as @ExecuteOn(TaskExecutors.BLOCKING), or use the reactive HTTP client to resolve this bug. There is also a configuration option to disable this check if you are certain a blocking operation is fine here."
```

**Solution**
Apply https://docs.micronaut.io/latest/guide/index.html#atBlocking

### Environment Information

- JDK 21

### Micronaut Version

4.3.7

**Comment by dstepanov** [comment]
Please create an issue for the first case. I think we can support it. WDYT @yawkat ?

**Comment by dstepanov** [comment]
Case 2: Considering the Jackson serializers depend on the Databind we cannot support it, the solution is to use `@Serializable(using=FooSerializer.class)`

Case 3: Please provide an example of this case. Do you code block?
